### DISCUSSION TITLE
How to tune OPENTHREAD_CONFIG to 32 child per parent

### DISCUSSION THREAD

**Discussion by rretanubun** [openthread/openthread#9950] [Q&A] [answered]
## Seeking advice on our approach
We have a customer request to have a TBR or TME that can buffer 8 frames (one full MTU) per child plus 100 shared frames.  Which works out to `32 +13 = 45 MTUs (full 1280-byte IP packets)`

## The code offers these `CONFIG` options:
1. `#define OPENTHREAD_CONFIG_MESSAGE_BUFFER_SIZE (sizeof(void *) * 32) `// which is 128 bytes in a 32-bit system
2. `#define OPENTHREAD_CONFIG_NUM_MESSAGE_BUFFERS 256` // which we bump to `512`
3. `#define OPENTHREAD_CONFIG_DEFAULT_SED_BUFFER_SIZE 1280` // which we kept `as is`
4. `#define OPENTHREAD_CONFIG_DEFAULT_SED_DATAGRAM_COUNT 1` // which we change to `13`

### Why `OPENTHREAD_CONFIG_NUM_MESSAGE_BUFFERS` changed?
1. Each `#define OPENTHREAD_CONFIG_MESSAGE_BUFFER_SIZE (sizeof(void *) * 32)` means the message buffer is 4*32 = 128 bytes
2. to support `full 1280-byte IP packets` we need 10 `OPENTHREAD_CONFIG_MESSAGE_BUFFER_SIZE `
3. we need 45 MTUs (1280-bytes each) so we need 45*10 = 450 `OPENTHREAD_CONFIG_NUM_MESSAGE_BUFFERS` which is why we are changing to `512` (rounding up for wiggle room and to account for the `100 shared frames`)

### Why `OPENTHREAD_CONFIG_DEFAULT_SED_DATAGRAM_COUNT` changed?
1. The request is to buffer `8 frames (one full MTU) per child` each MTU is `1280-byte` and each `SED datagram is 106-octet` (see the info on[ OPENTHREAD_CONFIG_DEFAULT_SED_DATAGRAM_COUNT](https://github.com/openthread/openthread/blob/51ab865600d73c78eab0900d2790f56ef6631a20/src/core/config/misc.h#L436-L443), to contain 1280 we need 1280/106 = 12.07, so we round up to `13`.

## Questions
1. Does the math checks out or am I missing something?
2. Any other CONFIG option that we should consider/tune for the goal?
3. Once a message is moved to the `SED Buffer` does it not longer consume space in the `Message Buffer`?

Thanks for everyone's time.

**Comment by abtink** [comment]
> `to support full 1280-byte IP packets we need 10 OPENTHREAD_CONFIG_MESSAGE_BUFFER_SIZE ` (each 128 byte)

- Need to also consider the `Metadata` saved in `Message` 
- And the reserved space for headers.
- Would recommend to account `11` or more (instead of `10`) data buffer (of 128 bytes length) to contain an IPv6 message with `1280` payload.

---

> Once a message is moved to the SED Buffer does it not longer consume space in the Message Buffer?

No. `Message Buffers` are shared/used for all.

---

> that can buffer 8 frames (one full MTU) per child plus 100 shared frames. Which works out to 32 +13 = 45 MTUs (full 1280-byte IP packets)


- You want 11 `MESSAGE_BUFFER` per 1280 byte IPv6 message.
- If you want to support 8 IPv6 msg (max 1280 bytes) per child, it becomes `8 * 11 = 88` buffers per child.


-----
You can mostly ignore `OPENTHREAD_CONFIG_DEFAULT_SED_BUFFER_SIZE`, `OPENTHREAD_CONFIG_DEFAULT_SED_DATAGRAM_COUNT` configs. 
These are only included in the emitted `Connectivity` TLV from device (they wont really adjust/allocate the buffers).

---

Note that IPv6 `Message` instances are placed in the message queue for the SED child. 6LoWPAN (fragment) frames are constructed when the stack transmits to the child. Additionally, a single IPv6 `Message` instance is used for multicast transmissions destined for multiple children.

**Comment by rretanubun** [comment]
@abtink : Thanks for the thoughtful reply. Double checking my understanding on some of the replies:

> You want 11 MESSAGE_BUFFER per 1280 byte IPv6 message.
> If you want to support 8 IPv6 msg (max 1280 bytes) per child, it becomes 8 * 11 = 88 buffers per child

This mean `(88 buffers per child * 32 child) + (13 MTU (for 100 shared frames) * 11 buffers-per-MTU) = 2959` for `OPENTHREAD_CONFIG_MESSAGE_BUFFER_SIZE ` if we want 8 IPv6 msg (max 1280 bytes) per child.

If we want 1 IPv6 msg (max 1280 bytes) per child, then it is `11 * (32 child +13 for-shared-frames) = 495` for `OPENTHREAD_CONFIG_MESSAGE_BUFFER_SIZE`
Correct?

Just curious, is there an advantage to supporting `8` (vs `1`) IPv6 msg per child?

> You can mostly ignore OPENTHREAD_CONFIG_DEFAULT_SED_BUFFER_SIZE, OPENTHREAD_CONFIG_DEFAULT_SED_DATAGRAM_COUNT configs.
> These are only included in the emitted Connectivity TLV from device (they wont really adjust/allocate the buffers).

Just making sure I follow, setting these higher does not increase the FTD/parent's capacity for processing/buffering SED specific messages for the children until they woke up?

**Reply by abtink** [reply, marked as answer]
> f we want 1 IPv6 msg (max 1280 bytes) per child, then it is 11 * (32 child +13 for-shared-frames) = 495 for OPENTHREAD_CONFIG_MESSAGE_BUFFER_SIZE
Correct?

Sounds correct to me.

> Just making sure I follow, setting these higher does not increase the FTD/parent's capacity for processing/buffering SED specific messages for the children until they woke up?

Correct.
